#Построить отказоустойчивый кластер PostgreSQL в облаках на базе Патрони.



#-- проверим, что c etcd
house@etcd1:~$ hostname; ps -aef | grep etcd | grep -v grep
etcd1
etcd        2661       1  0 19:34 ?        00:00:00 /usr/bin/etcd

house@etcd2:~$ hostname; ps -aef | grep etcd | grep -v grep
etcd2
etcd        2768       1  0 19:35 ?        00:00:00 /usr/bin/etcd

house@etcd3:~$ hostname; ps -aef | grep etcd | grep -v grep
etcd3
etcd        2124       1  0 19:35 ?        00:00:00 /usr/bin/etcd

#Остановил етсд на 3 хостах 
sudo systemctl stop etcd


#Выдал права 
house@etcd1:/var/lib/etcd$ sudo chmod 777 /var/lib/etcd/

#прописал ДНС

#--на 3 хостах добавим в файлы с конфигами /etc/default/etcd:

cat > temp.cfg << EOF 
ETCD_NAME="$(hostname)"
ETCD_LISTEN_CLIENT_URLS="http://0.0.0.0:2379"
ETCD_ADVERTISE_CLIENT_URLS="http://$(hostname):2379"
ETCD_LISTEN_PEER_URLS="http://0.0.0.0:2380"
ETCD_INITIAL_ADVERTISE_PEER_URLS="http://$(hostname):2380"
ETCD_INITIAL_CLUSTER_TOKEN="PatroniCluster"
ETCD_INITIAL_CLUSTER="etcd1=http://etcd1:2380,etcd2=http://etcd2:2380,etcd3=http://etcd3:2380"
ETCD_INITIAL_CLUSTER_STATE="new"
ETCD_DATA_DIR="/var/lib/etcd"
EOF
cat temp.cfg | sudo tee -a /etc/default/etcd


#-- проверка автозагрузки
house@etcd1:~$ systemctl is-enabled etcd
enabled

# кластера здороы
house@etcd1:~$ etcdctl cluster-health
member 9a1f33941721f94d is healthy: got healthy result from http://etcd1:2379
member 9df0146dd9068bd2 is healthy: got healthy result from http://etcd3:2379
member f2aeb69aaf7ffcbf is healthy: got healthy result from http://etcd2:2379
cluster is healthy



#--Изменил конфиг - патрони развернул, уронил 1 ноду - лидер переключился на 3 ноду.
house@pgsql1:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member |     Host    |   Role  |  State  | TL | Lag in MB |
+ Cluster: patroni1 (7137356545074075286) -+----+-----------+
| pgsql1 | 10.129.0.22 | Leader  | running |  2 |           |
| pgsql2 | 10.129.0.34 | Replica | running |  2 |        15 |
| pgsql3 | 10.129.0.23 | Replica | running |  2 |         0 |
+--------+-------------+---------+---------+----+-----------+

house@pgsql1:~$ sudo systemctl stop patroni

house@pgsql1:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member |     Host    |   Role  |  State  | TL | Lag in MB |
+ Cluster: patroni1 (7137356545074075286) -+----+-----------+
| pgsql1 | 10.129.0.22 | Replica | stopped |    |   unknown |
| pgsql2 | 10.129.0.34 | Replica | running |  2 |        31 |
| pgsql3 | 10.129.0.23 | Leader  | running |  3 |           |
+--------+-------------+---------+---------+----+-----------+

#-- установим pg_bouncer на каждом хосте с Патрони

sudo apt install -y pgbouncer

#зальем конфиг на 3 ноды
cat > temp3.cfg << EOF 
[databases]
otus = host=127.0.0.1 port=5432 dbname=otus 
[pgbouncer]
logfile = /var/log/postgresql/pgbouncer.log
pidfile = /var/run/postgresql/pgbouncer.pid
listen_addr = *
listen_port = 6432
auth_type = md5
auth_file = /etc/pgbouncer/userlist.txt
admin_users = admindb
EOF
cat temp3.cfg | sudo tee -a /etc/pgbouncer/pgbouncer.ini


#зальем конфиг на 3 ноды
cat > temp4.cfg << EOF 
"admindb" "d9cfab6a2f1a0eb0c037e605cd578025"
EOF
cat temp4.cfg | sudo tee -a /etc/pgbouncer/userlist.txt

#pgbouncer статус
house@postgres:~$ sudo systemctl status pgbouncer
● pgbouncer.service - pgbouncer PostgreSQL connection pooler
     Loaded: loaded (/lib/systemd/system/pgbouncer.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2022-05-26 20:51:13 UTC; 21min ago
   Main PID: 2175 (pgbouncer)
     Status: "stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 μs, query 0 μs, wait 0 μs"
      Tasks: 2 (limit: 2311)
     Memory: 1.2M
     CGroup: /system.slice/pgbouncer.service
             └─2175 /usr/sbin/pgbouncer /etc/pgbouncer/pgbouncer.ini

May 26 21:03:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:04:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:05:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:06:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:07:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:08:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:09:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:10:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:11:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
May 26 21:12:13 postgres pgbouncer[2175]: stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us

#--настроил на на 3 ноды userlist 
house@postgres:~$sudo nano /etc/pgbouncer/userlist.txt
"postgres"  "SCRAM-SHA-256$4096:5QzE5QKo/Y0sUSKxIIhmBA==$mViw6DFyxLkcRRDyft3UOoJ/JGw5zRyvpOJnkt+EtmI=:8gi9vxsVajxpmcVZb3gqjnecPeXoel/eBqk3pEbbX7k="


house@postgres:~$ sudo nano /etc/pgbouncer/pgbouncer.ini
admin_users = postgres


#--На след день 1 и 2 ноды не стартанули - перезапуск патрони и ВМ не помогает
house@pgsql1:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+--------------+----+-----------+
| Member |     Host    |   Role  |    State     | TL | Lag in MB |
+ Cluster: patroni1 (7137356545074075286) ------+----+-----------+
| pgsql1 | 10.129.0.22 | Replica | starting     |    |   unknown |
| pgsql2 | 10.129.0.34 | Replica | start failed |    |   unknown |
| pgsql3 | 10.129.0.23 | Leader  | running      |  4 |           |
+--------+-------------+---------+--------------+----+-----------+


house@pgsql1:~$ sudo journalctl -eu patroni.service
Aug 31 13:59:13 pgsql1 patroni[651]: 2022-08-31 13:59:13,202 INFO: establishing a new patroni connection to the postgres cluster
Aug 31 13:59:13 pgsql1 patroni[2054]: 2022-08-31 13:59:13.204 UTC [2054] FATAL:  the database system is starting up
Aug 31 13:59:13 pgsql1 patroni[651]: 2022-08-31 13:59:13,204 WARNING: Retry got exception: 'connection problems'
Aug 31 13:59:13 pgsql1 patroni[651]: 2022-08-31 13:59:13,205 WARNING: Failed to determine PostgreSQL state from the connection, falling bac>
Aug 31 13:59:13 pgsql1 patroni[651]: 2022-08-31 13:59:13,209 INFO: no action. I am (pgsql1), a secondary, and following a leader (pgsql3)
Aug 31 13:59:22 pgsql1 patroni[2056]: 2022-08-31 13:59:22.698 UTC [2056] FATAL:  the database system is starting up
Aug 31 13:59:22 pgsql1 patroni[2055]: localhost:5432 - rejecting connections
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,700 INFO: Lock owner: pgsql3; I am pgsql1
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,700 INFO: Still starting up as a standby.
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,701 INFO: Lock owner: pgsql3; I am pgsql1
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,701 INFO: establishing a new patroni connection to the postgres cluster
Aug 31 13:59:22 pgsql1 patroni[2058]: 2022-08-31 13:59:22.702 UTC [2058] FATAL:  the database system is starting up
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,953 INFO: establishing a new patroni connection to the postgres cluster
Aug 31 13:59:22 pgsql1 patroni[2059]: 2022-08-31 13:59:22.954 UTC [2059] FATAL:  the database system is starting up
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,955 WARNING: Retry got exception: 'connection problems'
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,955 WARNING: Failed to determine PostgreSQL state from the connection, falling bac>
Aug 31 13:59:22 pgsql1 patroni[651]: 2022-08-31 13:59:22,959 INFO: no action. I am (pgsql1), a secondary, and following a leader (pgsql3)
Aug 31 13:59:32 pgsql1 patroni[2132]: 2022-08-31 13:59:32.698 UTC [2132] FATAL:  the database system is starting up
Aug 31 13:59:32 pgsql1 patroni[2131]: localhost:5432 - rejecting connections
Aug 31 13:59:32 pgsql1 patroni[651]: 2022-08-31 13:59:32,699 INFO: Lock owner: pgsql3; I am pgsql1
Aug 31 13:59:32 pgsql1 patroni[651]: 2022-08-31 13:59:32,699 INFO: Still starting up as a standby.
Aug 31 13:59:32 pgsql1 patroni[651]: 2022-08-31 13:59:32,700 INFO: Lock owner: pgsql3; I am pgsql1
Aug 31 13:59:32 pgsql1 patroni[651]: 2022-08-31 13:59:32,700 INFO: establishing a new patroni connection to the postgres cluster
Aug 31 13:59:32 pgsql1 patroni[2133]: 2022-08-31 13:59:32.701 UTC [2133] FATAL:  the database system is starting up
Aug 31 13:59:32 pgsql1 patroni[651]: 2022-08-31 13:59:32,992 INFO: establishing a new patroni connection to the postgres cluster
Aug 31 13:59:32 pgsql1 patroni[2134]: 2022-08-31 13:59:32.994 UTC [2134] FATAL:  the database system is starting up
Aug 31 13:59:32 pgsql1 patroni[651]: 2022-08-31 13:59:32,994 WARNING: Retry got exception: 'connection problems'
Aug 31 13:59:32 pgsql1 patroni[651]: 2022-08-31 13:59:32,994 WARNING: Failed to determine PostgreSQL state from the connection, falling bac>
Aug 31 13:59:33 pgsql1 patroni[651]: 2022-08-31 13:59:33,004 INFO: no action. I am (pgsql1), a secondary, and following a leader (pgsql3)



house@pgsql2:~$ sudo journalctl -eu patroni.service
Aug 31 13:58:42 pgsql2 patroni[650]:   Maximum size of a TOAST chunk: 1996
Aug 31 13:58:42 pgsql2 patroni[650]:   Size of a large-object chunk: 2048
Aug 31 13:58:42 pgsql2 patroni[650]:   Date/time type storage: 64-bit integers
Aug 31 13:58:42 pgsql2 patroni[650]:   Float8 argument passing: by value
Aug 31 13:58:42 pgsql2 patroni[650]:   Data page checksum version: 1
Aug 31 13:58:42 pgsql2 patroni[650]:   Mock authentication nonce: e2202fe740a3a26add922b57557f1d11069922f438b0d3b2cb54f1bd1d0bbcc2
Aug 31 13:58:42 pgsql2 patroni[650]: 2022-08-31 13:58:42,689 INFO: Lock owner: pgsql3; I am pgsql2
Aug 31 13:58:42 pgsql2 patroni[650]: 2022-08-31 13:58:42,695 INFO: Local timeline=2 lsn=0/40000A0
Aug 31 13:58:42 pgsql2 patroni[650]: 2022-08-31 13:58:42,709 INFO: master_timeline=5
Aug 31 13:58:42 pgsql2 patroni[650]: 2022-08-31 13:58:42,709 INFO: master: history=1        0/1953E38        no recovery target specified
Aug 31 13:58:42 pgsql2 patroni[650]: 2        0/60000A0        no recovery target specified
Aug 31 13:58:42 pgsql2 patroni[650]: 3        0/6000230        no recovery target specified
Aug 31 13:58:42 pgsql2 patroni[650]: 4        0/60003C0        no recovery target specified
Aug 31 13:58:42 pgsql2 patroni[650]: 2022-08-31 13:58:42,709 INFO: Lock owner: pgsql3; I am pgsql2
Aug 31 13:58:42 pgsql2 patroni[650]: 2022-08-31 13:58:42,713 INFO: starting as a secondary
Aug 31 13:58:42 pgsql2 patroni[650]: 2022-08-31 13:58:42,856 INFO: postmaster pid=2139
Aug 31 13:58:42 pgsql2 patroni[2140]: localhost:5432 - no response
Aug 31 13:58:42 pgsql2 patroni[2139]: 2022-08-31 13:58:42.864 UTC [2139] LOG:  starting PostgreSQL 14.5 (Ubuntu 14.5-1.pgdg22.04+1) on x86_>
Aug 31 13:58:42 pgsql2 patroni[2139]: 2022-08-31 13:58:42.864 UTC [2139] LOG:  listening on IPv4 address "127.0.0.1", port 5432
Aug 31 13:58:42 pgsql2 patroni[2139]: 2022-08-31 13:58:42.868 UTC [2139] LOG:  listening on IPv4 address "10.129.0.34", port 5432
Aug 31 13:58:42 pgsql2 patroni[2139]: 2022-08-31 13:58:42.871 UTC [2139] LOG:  listening on Unix socket "./.s.PGSQL.5432"
Aug 31 13:58:42 pgsql2 patroni[2141]: 2022-08-31 13:58:42.877 UTC [2141] LOG:  database system was shut down in recovery at 2022-08-30 19:1>
Aug 31 13:58:42 pgsql2 patroni[2141]: 2022-08-31 13:58:42.877 UTC [2141] LOG:  entering standby mode
Aug 31 13:58:42 pgsql2 patroni[2141]: 2022-08-31 13:58:42.877 UTC [2141] LOG:  invalid resource manager ID in primary checkpoint record
Aug 31 13:58:42 pgsql2 patroni[2141]: 2022-08-31 13:58:42.877 UTC [2141] PANIC:  could not locate a valid checkpoint record
Aug 31 13:58:42 pgsql2 patroni[2139]: 2022-08-31 13:58:42.987 UTC [2139] LOG:  startup process (PID 2141) was terminated by signal 6: Abort>
Aug 31 13:58:42 pgsql2 patroni[2139]: 2022-08-31 13:58:42.987 UTC [2139] LOG:  aborting startup due to startup process failure
Aug 31 13:58:42 pgsql2 patroni[2139]: 2022-08-31 13:58:42.988 UTC [2139] LOG:  database system is shut down
Aug 31 13:58:43 pgsql2 patroni[650]: 2022-08-31 13:58:43,864 ERROR: postmaster is not running


house@pgsql3:~$ sudo journalctl -eu patroni.service
Aug 31 13:53:02 pgsql3 patroni[650]: 2022-08-31 13:53:02,682 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:53:12 pgsql3 patroni[650]: 2022-08-31 13:53:12,687 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:53:22 pgsql3 patroni[650]: 2022-08-31 13:53:22,687 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:53:32 pgsql3 patroni[650]: 2022-08-31 13:53:32,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:53:42 pgsql3 patroni[650]: 2022-08-31 13:53:42,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:53:52 pgsql3 patroni[650]: 2022-08-31 13:53:52,682 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:54:02 pgsql3 patroni[650]: 2022-08-31 13:54:02,681 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:54:12 pgsql3 patroni[650]: 2022-08-31 13:54:12,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:54:22 pgsql3 patroni[650]: 2022-08-31 13:54:22,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:54:32 pgsql3 patroni[650]: 2022-08-31 13:54:32,678 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:54:42 pgsql3 patroni[650]: 2022-08-31 13:54:42,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:54:52 pgsql3 patroni[650]: 2022-08-31 13:54:52,680 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:55:02 pgsql3 patroni[650]: 2022-08-31 13:55:02,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:55:12 pgsql3 patroni[650]: 2022-08-31 13:55:12,680 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:55:22 pgsql3 patroni[650]: 2022-08-31 13:55:22,682 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:55:32 pgsql3 patroni[650]: 2022-08-31 13:55:32,678 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:55:42 pgsql3 patroni[650]: 2022-08-31 13:55:42,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:55:52 pgsql3 patroni[650]: 2022-08-31 13:55:52,678 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:56:02 pgsql3 patroni[650]: 2022-08-31 13:56:02,681 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:56:12 pgsql3 patroni[650]: 2022-08-31 13:56:12,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:56:22 pgsql3 patroni[650]: 2022-08-31 13:56:22,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:56:32 pgsql3 patroni[650]: 2022-08-31 13:56:32,681 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:56:42 pgsql3 patroni[650]: 2022-08-31 13:56:42,679 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:56:52 pgsql3 patroni[650]: 2022-08-31 13:56:52,680 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:57:02 pgsql3 patroni[650]: 2022-08-31 13:57:02,682 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:57:12 pgsql3 patroni[650]: 2022-08-31 13:57:12,680 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:57:22 pgsql3 patroni[650]: 2022-08-31 13:57:22,681 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:57:32 pgsql3 patroni[650]: 2022-08-31 13:57:32,681 INFO: no action. I am (pgsql3), the leader with the lock
Aug 31 13:57:42 pgsql3 patroni[650]: 2022-08-31 13:57:42,678 INFO: no action. I am (pgsql3), the leader with the lock






----------------------------------------------------------------------------------------------------------------------






----Удалил кластеры с патрони - создал новые.
----etcd оставил старые

house@etcd1:~$ etcdctl cluster-health
member 9a1f33941721f94d is healthy: got healthy result from http://etcd1.ru-central1.internal:2379
member 9df0146dd9068bd2 is healthy: got healthy result from http://etcd3.ru-central1.internal:2379
member f2aeb69aaf7ffcbf is healthy: got healthy result from http://etcd2.ru-central1.internal:2379
cluster is healthy



----Проверил пинги

house@etcd1:~$ ping pgsql1
PING pgsql1.ru-central1.internal (10.129.0.24) 56(84) bytes of data.
64 bytes from pgsql1.ru-central1.internal (10.129.0.24): icmp_seq=1 ttl=63 time=0.387 ms

house@pgsql1:~$ ping etcd1
PING etcd1.ru-central1.internal (10.129.0.21) 56(84) bytes of data.
64 bytes from etcd1.ru-central1.internal (10.129.0.21): icmp_seq=1 ttl=63 time=1.01 ms



#-- установка: постгрес на 3 ноды
sudo apt update && sudo apt upgrade -y -q && echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main" | sudo tee -a /etc/apt/sources.list.d/pgdg.list && wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | sudo apt-key add - && sudo apt-get update && sudo apt -y install postgresql-14


#----убедимся, что кластера Постгреса стартовали
house@pgsql1:~$ hostname; pg_lsclusters
pgsql1
Ver Cluster Port Status Owner    Data directory              Log file
14  main    5432 online postgres /var/lib/postgresql/14/main /var/log/postgresql/postgresql-14-main.log


#----ставим питон на 1 ноде
sudo apt-get install -y python3 python3-pip git mc
sudo pip3 install psycopg2-binary 

#---- после установки ПО останавливаем и удаляем экземлпяр постгреса который запускается по-умолчанию:
house@pgsql1:~$ sudo systemctl stop postgresql@14-main
house@pgsql1:~$ sudo -u postgres pg_dropcluster 14 main
Warning: systemd was not informed about the removed cluster yet. Operations like "service postgresql start" might fail. To fix, run:
  sudo systemctl daemon-reload

#----убеждемся что их нет
house@pgsql1:~$ pg_lsclusters
Ver Cluster Port Status Owner Data directory Log file

#----патрони 
sudo pip3 install patroni[etcd]

#----делаем симлинк
sudo ln -s /usr/local/bin/patroni /bin/patroni

#---- закидываем конфиги

----1
cat > temp.cfg << EOF 
[Unit]
Description=High availability PostgreSQL Cluster
After=syslog.target network.target
[Service]
Type=simple
User=postgres
Group=postgres
ExecStart=/usr/local/bin/patroni /etc/patroni.yml
KillMode=process
TimeoutSec=30
Restart=no
[Install]
WantedBy=multi-user.target
EOF
cat temp.cfg | sudo tee -a /etc/systemd/system/patroni.service


----2
cat > temp2.cfg << EOF 
scope: patroni2
name: $(hostname)
restapi:
  listen: $(hostname -I | tr -d " "):8008
  connect_address: $(hostname -I | tr -d " "):8008
etcd:
  hosts: etcd1.ru-central1.internal:2379,etcd2.ru-central1.internal:2379,etcd3.ru-central1.internal:2379
bootstrap:
  dcs:
    ttl: 30
    loop_wait: 10
    retry_timeout: 10
    maximum_lag_on_failover: 1048576
    postgresql:
      use_pg_rewind: true
      parameters:
  initdb: 
  - encoding: UTF8
  - data-checksums
  pg_hba: 
  - host replication replicator 10.0.0.0/8 md5
  - host all all 10.0.0.0/8 md5
  users:
    admin:
      password: admin_321
      options:
        - createrole
        - createdb
postgresql:
  listen: 127.0.0.1, $(hostname -I | tr -d " "):5432
  connect_address: $(hostname -I | tr -d " "):5432
  data_dir: /var/lib/postgresql/14/main
  bin_dir: /usr/lib/postgresql/14/bin
  pgpass: /tmp/pgpass0
  authentication:
    replication:
      username: replicator
      password: rep-pass_321
    superuser:
      username: postgres
      password: zalando_321
    rewind:  
      username: rewind_user
      password: rewind_password_321
  parameters:
    unix_socket_directories: '.'
tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false
EOF
cat temp2.cfg | sudo tee -a /etc/patroni.yml



#----бутстрапим команда не заканчивается выдает: not healthy enough for leader race
#----Прерываю, в итоге все ок 

house@pgsql1:~$ sudo -u postgres patroni /etc/patroni.yml
2022-09-02 14:38:20,021 INFO: Selected new etcd server http://etcd2.ru-central1.internal:2379
2022-09-02 14:38:20,026 INFO: No PostgreSQL configuration items changed, nothing to reload.
2022-09-02 14:38:20,033 INFO: Lock owner: None; I am pgsql1
2022-09-02 14:38:20,038 INFO: trying to bootstrap a new cluster
could not change directory to "/home/house": Permission denied
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.UTF-8".
The default text search configuration will be set to "english".

Data page checksums are enabled.

creating directory /var/lib/postgresql/14/main ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... Etc/UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

initdb: warning: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    /usr/lib/postgresql/14/bin/pg_ctl -D /var/lib/postgresql/14/main -l logfile start

could not change directory to "/home/house": Permission denied
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 225, in prepare
    os.chdir(data['dir'])
PermissionError: [Errno 13] Permission denied: '/home/house'
2022-09-02 14:38:30,030 INFO: Lock owner: None; I am pgsql1
2022-09-02 14:38:30,030 INFO: not healthy enough for leader race
2022-09-02 14:38:30,037 INFO: bootstrap in progress
2022-09-02 14:38:40,030 INFO: Lock owner: None; I am pgsql1
...
...
...
...
2022-09-02 14:39:50,030 INFO: not healthy enough for leader race
2022-09-02 14:39:50,034 INFO: bootstrap in progress
2022-09-02 14:40:00,030 INFO: Lock owner: None; I am pgsql1
2022-09-02 14:40:00,030 INFO: not healthy enough for leader race
2022-09-02 14:40:00,034 INFO: bootstrap in progress
^C^CException ignored in: <module 'threading' from '/usr/lib/python3.10/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1560, in _shutdown
    lock.acquire()
KeyboardInterrupt:


#----Ставим автозапуск и запускаем
house@pgsql1:~$ sudo systemctl is-enabled patroni
disabled

house@pgsql1:~$ sudo systemctl enable patroni
Created symlink /etc/systemd/system/multi-user.target.wants/patroni.service → /etc/systemd/system/patroni.service.

house@pgsql1:~$ sudo systemctl status patroni
○ patroni.service - High availability PostgreSQL Cluster
     Loaded: loaded (/etc/systemd/system/patroni.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
	 
house@pgsql1:~$ sudo systemctl start patroni

house@pgsql1:~$ sudo systemctl status patroni
● patroni.service - High availability PostgreSQL Cluster
     Loaded: loaded (/etc/systemd/system/patroni.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2022-09-02 14:41:36 UTC; 4s ago
   Main PID: 18135 (patroni)
      Tasks: 13 (limit: 2236)
     Memory: 65.2M
        CPU: 420ms
     CGroup: /system.slice/patroni.service
             ├─18135 /usr/bin/python3 /usr/local/bin/patroni /etc/patroni.yml
             ├─18148 /usr/lib/postgresql/14/bin/postgres -D /var/lib/postgresql/14/main --config-file=/var/lib/postgresql/14/main/postgresql.c>
             ├─18151 "postgres: patroni2: checkpointer " "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" ">
             ├─18152 "postgres: patroni2: background writer " "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "">
             ├─18153 "postgres: patroni2: stats collector " "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" ">
             ├─18158 "postgres: patroni2: postgres postgres 127.0.0.1(57324) idle" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "">
             ├─18166 "postgres: patroni2: walwriter " "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" ">
             ├─18167 "postgres: patroni2: autovacuum launcher " "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" >
             └─18168 "postgres: patroni2: logical replication launcher " "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" >

Sep 02 14:41:37 pgsql1 patroni[18135]: connection to server at "localhost" (127.0.0.1), port 5432 failed: FATAL:  role "replicator" does not e>
Sep 02 14:41:37 pgsql1 patroni[18135]: 2022-09-02 14:41:37,846 INFO: promoted self to leader by acquiring session lock
Sep 02 14:41:37 pgsql1 patroni[18161]: server promoting
Sep 02 14:41:37 pgsql1 patroni[18150]: 2022-09-02 14:41:37.847 UTC [18150] LOG:  received promote request
Sep 02 14:41:37 pgsql1 patroni[18150]: 2022-09-02 14:41:37.847 UTC [18150] LOG:  redo is not required
Sep 02 14:41:37 pgsql1 patroni[18135]: 2022-09-02 14:41:37,848 INFO: cleared rewind state after becoming the leader
Sep 02 14:41:37 pgsql1 patroni[18150]: 2022-09-02 14:41:37.853 UTC [18150] LOG:  selected new timeline ID: 2
Sep 02 14:41:39 pgsql1 patroni[18150]: 2022-09-02 14:41:39.054 UTC [18150] LOG:  archive recovery complete
Sep 02 14:41:39 pgsql1 patroni[18148]: 2022-09-02 14:41:39.081 UTC [18148] LOG:  database system is ready to accept connections
Sep 02 14:41:39 pgsql1 patroni[18135]: 2022-09-02 14:41:39,943 INFO: no action. I am (pgsql1), the leader with the lock
lines 1-28/28 (END)


house@pgsql1:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+--------+---------+----+-----------+
| Member | Host        | Role   | State   | TL | Lag in MB |
+ Cluster: patroni2 (7138791844683556511) +----+-----------+
| pgsql1 | 10.129.0.24 | Leader | running |  2 |           |
+--------+-------------+--------+---------+----+-----------+




#----На 2 и 3 нодах бутстрап заканчивался другой ошибкой

house@pgsql3:~$ sudo -u postgres patroni /etc/patroni.yml
2022-09-02 15:32:21,996 INFO: Selected new etcd server http://etcd3.ru-central1.internal:2379
2022-09-02 15:32:22,002 INFO: No PostgreSQL configuration items changed, nothing to reload.
2022-09-02 15:32:22,008 INFO: Lock owner: pgsql1; I am pgsql3
2022-09-02 15:32:22,011 INFO: trying to bootstrap from leader 'pgsql1'
could not change directory to "/home/house": Permission denied
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authentication failed for user "replicator"
password retrieved from file "/tmp/pgpass0"
2022-09-02 15:32:22,029 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2022-09-02 15:32:22,029 WARNING: Trying again in 5 seconds
could not change directory to "/home/house": Permission denied
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authentication failed for user "replicator"
password retrieved from file "/tmp/pgpass0"
2022-09-02 15:32:27,049 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2022-09-02 15:32:27,049 ERROR: failed to bootstrap from leader 'pgsql1'
2022-09-02 15:32:27,049 INFO: Removing data directory: /var/lib/postgresql/14/main
2022-09-02 15:32:32,006 INFO: Lock owner: pgsql1; I am pgsql3
2022-09-02 15:32:32,012 INFO: trying to bootstrap from leader 'pgsql1'
could not change directory to "/home/house": Permission denied
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authentication failed for user "replicator"
password retrieved from file "/tmp/pgpass0"
2022-09-02 15:32:32,029 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2022-09-02 15:32:32,029 WARNING: Trying again in 5 seconds
could not change directory to "/home/house": Permission denied
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authentication failed for user "replicator"
password retrieved from file "/tmp/pgpass0"
2022-09-02 15:32:37,050 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2022-09-02 15:32:37,050 ERROR: failed to bootstrap from leader 'pgsql1'
2022-09-02 15:32:37,050 INFO: Removing data directory: /var/lib/postgresql/14/main
2022-09-02 15:32:42,006 INFO: Lock owner: pgsql1; I am pgsql3
2022-09-02 15:32:42,010 INFO: trying to bootstrap from leader 'pgsql1'
could not change directory to "/home/house": Permission denied
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authentication failed for user "replicator"
password retrieved from file "/tmp/pgpass0"
2022-09-02 15:32:42,026 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2022-09-02 15:32:42,026 WARNING: Trying again in 5 seconds
^Ccould not change directory to "/home/house": Permission denied
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authentication failed for user "replicator"
password retrieved from file "/tmp/pgpass0"


#----Заупстил патрони 
house@pgsql3:~$ sudo systemctl is-enabled patroni
disabled

house@pgsql3:~$ sudo systemctl enable patroni
Created symlink /etc/systemd/system/multi-user.target.wants/patroni.service → /etc/systemd/system/patroni.service.

house@pgsql3:~$ sudo systemctl status patroni
○ patroni.service - High availability PostgreSQL Cluster
     Loaded: loaded (/etc/systemd/system/patroni.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
	 
house@pgsql3:~$ sudo systemctl start patroni


#----Проверил статус 
house@pgsql3:~$ sudo systemctl status patroni
● patroni.service - High availability PostgreSQL Cluster
     Loaded: loaded (/etc/systemd/system/patroni.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2022-09-02 15:33:11 UTC; 5s ago
   Main PID: 18312 (patroni)
      Tasks: 5 (limit: 2236)
     Memory: 27.6M
        CPU: 249ms
     CGroup: /system.slice/patroni.service
             └─18312 /usr/bin/python3 /usr/local/bin/patroni /etc/patroni.yml

Sep 02 15:33:11 pgsql3 patroni[18312]: 2022-09-02 15:33:11,766 INFO: trying to bootstrap from leader 'pgsql1'
Sep 02 15:33:11 pgsql3 patroni[18318]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 15:33:11 pgsql3 patroni[18318]: password retrieved from file "/tmp/pgpass0"
Sep 02 15:33:11 pgsql3 patroni[18312]: 2022-09-02 15:33:11,783 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 15:33:11 pgsql3 patroni[18312]: 2022-09-02 15:33:11,783 WARNING: Trying again in 5 seconds
Sep 02 15:33:16 pgsql3 patroni[18319]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 15:33:16 pgsql3 patroni[18319]: password retrieved from file "/tmp/pgpass0"
Sep 02 15:33:16 pgsql3 patroni[18312]: 2022-09-02 15:33:16,803 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 15:33:16 pgsql3 patroni[18312]: 2022-09-02 15:33:16,803 ERROR: failed to bootstrap from leader 'pgsql1'
Sep 02 15:33:16 pgsql3 patroni[18312]: 2022-09-02 15:33:16,804 INFO: Removing data directory: /var/lib/postgresql/14/main


#---- Проверил пинг 
house@pgsql3:~$ ping pgsql1
PING pgsql1.ru-central1.internal (10.129.0.24) 56(84) bytes of data.
64 bytes from pgsql1.ru-central1.internal (10.129.0.24): icmp_seq=1 ttl=63 time=0.854 ms
64 bytes from pgsql1.ru-central1.internal (10.129.0.24): icmp_seq=2 ttl=63 time=0.414 ms
64 bytes from pgsql1.ru-central1.internal (10.129.0.24): icmp_seq=3 ttl=63 time=0.387 ms
^C
--- pgsql1.ru-central1.internal ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2011ms
rtt min/avg/max/mdev = 0.387/0.551/0.854/0.214 ms


#----
house@pgsql1:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni2 (7138791844683556511) -+----+-----------+
| pgsql1 | 10.129.0.24 | Leader  | running |  2 |           |
| pgsql2 | 10.129.0.19 | Replica | stopped |    |   unknown |
| pgsql3 | 10.129.0.15 | Replica | stopped |    |   unknown |
+--------+-------------+---------+---------+----+-----------+


house@pgsql2:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni2 (7138791844683556511) -+----+-----------+
| pgsql1 | 10.129.0.24 | Leader  | running |  2 |           |
| pgsql2 | 10.129.0.19 | Replica | stopped |    |   unknown |
| pgsql3 | 10.129.0.15 | Replica | stopped |    |   unknown |
+--------+-------------+---------+---------+----+-----------+


#----Проверил журнал - не может подключиться к лидеру   trying to bootstrap from leader 'pgsql1'
house@pgsql2:~$ sudo journalctl -eu patroni.service
Sep 02 17:05:13 pgsql2 patroni[18027]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 17:05:13 pgsql2 patroni[18027]: password retrieved from file "/tmp/pgpass0"
Sep 02 17:05:13 pgsql2 patroni[18004]: 2022-09-02 17:05:13,198 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 17:05:13 pgsql2 patroni[18004]: 2022-09-02 17:05:13,198 ERROR: failed to bootstrap from leader 'pgsql1'
Sep 02 17:05:13 pgsql2 patroni[18004]: 2022-09-02 17:05:13,198 INFO: Removing data directory: /var/lib/postgresql/14/main
Sep 02 17:05:18 pgsql2 patroni[18004]: 2022-09-02 17:05:18,150 INFO: Lock owner: pgsql1; I am pgsql2
Sep 02 17:05:18 pgsql2 patroni[18004]: 2022-09-02 17:05:18,154 INFO: trying to bootstrap from leader 'pgsql1'
Sep 02 17:05:18 pgsql2 patroni[18029]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 17:05:18 pgsql2 patroni[18029]: password retrieved from file "/tmp/pgpass0"
Sep 02 17:05:18 pgsql2 patroni[18004]: 2022-09-02 17:05:18,171 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 17:05:18 pgsql2 patroni[18004]: 2022-09-02 17:05:18,171 WARNING: Trying again in 5 seconds
Sep 02 17:05:23 pgsql2 patroni[18030]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 17:05:23 pgsql2 patroni[18030]: password retrieved from file "/tmp/pgpass0"
Sep 02 17:05:23 pgsql2 patroni[18004]: 2022-09-02 17:05:23,192 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 17:05:23 pgsql2 patroni[18004]: 2022-09-02 17:05:23,192 ERROR: failed to bootstrap from leader 'pgsql1'
Sep 02 17:05:23 pgsql2 patroni[18004]: 2022-09-02 17:05:23,192 INFO: Removing data directory: /var/lib/postgresql/14/main
Sep 02 17:05:28 pgsql2 patroni[18004]: 2022-09-02 17:05:28,151 INFO: Lock owner: pgsql1; I am pgsql2
Sep 02 17:05:28 pgsql2 patroni[18004]: 2022-09-02 17:05:28,154 INFO: trying to bootstrap from leader 'pgsql1'
Sep 02 17:05:28 pgsql2 patroni[18032]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 17:05:28 pgsql2 patroni[18032]: password retrieved from file "/tmp/pgpass0"
Sep 02 17:05:28 pgsql2 patroni[18004]: 2022-09-02 17:05:28,171 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 17:05:28 pgsql2 patroni[18004]: 2022-09-02 17:05:28,171 WARNING: Trying again in 5 seconds
Sep 02 17:05:33 pgsql2 patroni[18033]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 17:05:33 pgsql2 patroni[18033]: password retrieved from file "/tmp/pgpass0"
Sep 02 17:05:33 pgsql2 patroni[18004]: 2022-09-02 17:05:33,192 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 17:05:33 pgsql2 patroni[18004]: 2022-09-02 17:05:33,193 ERROR: failed to bootstrap from leader 'pgsql1'
Sep 02 17:05:33 pgsql2 patroni[18004]: 2022-09-02 17:05:33,193 INFO: Removing data directory: /var/lib/postgresql/14/main
Sep 02 17:05:38 pgsql2 patroni[18004]: 2022-09-02 17:05:38,150 INFO: Lock owner: pgsql1; I am pgsql2
Sep 02 17:05:38 pgsql2 patroni[18004]: 2022-09-02 17:05:38,153 INFO: trying to bootstrap from leader 'pgsql1'
Sep 02 17:05:38 pgsql2 patroni[18035]: pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  password authent>
Sep 02 17:05:38 pgsql2 patroni[18035]: password retrieved from file "/tmp/pgpass0"
Sep 02 17:05:38 pgsql2 patroni[18004]: 2022-09-02 17:05:38,170 ERROR: Error when fetching backup: pg_basebackup exited with code=1
Sep 02 17:05:38 pgsql2 patroni[18004]: 2022-09-02 17:05:38,170 WARNING: Trying again in 5 seconds



#----ВМ 
ru-central1-b
10.129.0.21
158.160.11.168
etcd1

ru-central1-b
10.129.0.16
158.160.7.88
etcd2

ru-central1-b
10.129.0.3
158.160.6.46
etcd3

ru-central1-b
10.129.0.24
51.250.27.185
pgsql1

ru-central1-b
10.129.0.19
51.250.96.23
pgsql2

ru-central1-b
10.129.0.15
84.252.138.198
pgsql3


--------------------------------------------------------------------------------------------------------------------------


#----Создал пользователя:

house@pgsql1:~$ sudo -u postgres  psql -h localhost
could not change directory to "/home/house": Permission denied
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1))
Type "help" for help.

postgres=# create user replicator with password 'rep-pass_321';
CREATE ROLE
postgres=# \du
                                    List of roles
 Role name  |                         Attributes                         | Member of
------------+------------------------------------------------------------+-----------
 postgres   | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 replicator |                                                            | {}

#----выдал права, ушла ошибка ---- could not change directory to "/home/house": Permission denied 
house@pgsql1:~$ chmod 777 /home/house

house@pgsql1:~$ sudo -u postgres  psql -h localhost
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1))
Type "help" for help.

postgres=#


#----На второй ноде запустил бутсрап - ошибка с доступом /home/house": Permission denied
house@pgsql2:~$ sudo systemctl stop patroni

house@pgsql2:~$ sudo -u postgres patroni /etc/patroni.yml
2022-09-05 17:34:55,085 INFO: Selected new etcd server http://etcd3.ru-central1.internal:2379
2022-09-05 17:34:55,090 INFO: No PostgreSQL configuration items changed, nothing to reload.
2022-09-05 17:34:55,093 INFO: Lock owner: pgsql1; I am pgsql2
2022-09-05 17:34:55,099 INFO: trying to bootstrap from leader 'pgsql1'
could not change directory to "/home/house": Permission denied
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  must be superuser or replication role to start walsender
2022-09-05 17:34:55,119 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2022-09-05 17:34:55,120 WARNING: Trying again in 5 seconds


#----выдал права
house@pgsql2:~$ chmod 777 /home/house


house@pgsql2:~$ sudo systemctl start patroni
house@pgsql2:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni2 (7138791844683556511) -+----+-----------+
| pgsql1 | 10.129.0.24 | Leader  | running |  6 |           |
| pgsql2 | 10.129.0.19 | Replica | stopped |    |   unknown |
| pgsql3 | 10.129.0.15 | Replica | stopped |    |   unknown |
+--------+-------------+---------+---------+----+-----------+


house@pgsql2:~$ sudo systemctl stop patroni


#---- просит суперпользователь или роль репликации
house@pgsql2:~$ sudo -u postgres patroni /etc/patroni.yml
2022-09-05 17:37:31,899 INFO: Selected new etcd server http://etcd1.ru-central1.internal:2379
2022-09-05 17:37:31,905 INFO: No PostgreSQL configuration items changed, nothing to reload.
2022-09-05 17:37:31,908 INFO: Lock owner: pgsql1; I am pgsql2
2022-09-05 17:37:31,913 INFO: trying to bootstrap from leader 'pgsql1'
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  must be superuser or replication role to start walsender
2022-09-05 17:37:31,930 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2022-09-05 17:37:31,930 WARNING: Trying again in 5 seconds
pg_basebackup: error: connection to server at "10.129.0.24", port 5432 failed: FATAL:  must be superuser or replication role to start walsender
2022-09-05 17:37:36,952 ERROR: Error when fetching backup: pg_basebackup exited with code=1



#---- Решил перезагрузить ноды - из-за выдачи прав, похоже изменился authorized_keys  потерял доступ к 1 и 2 ноде

house@Igor:/mnt/c/Users/house$ ssh house@84.201.165.237
house@84.201.165.237: Permission denied (publickey).


#----Остался доступ только к 3ей, манипуляции выше не помогли.

house@pgsql3:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni2 (7138791844683556511) -+----+-----------+
| pgsql1 | 10.129.0.24 | Leader  | running |  8 |           |
| pgsql2 | 10.129.0.19 | Replica | stopped |    |   unknown |
| pgsql3 | 10.129.0.15 | Replica | stopped |    |   unknown |
+--------+-------------+---------+---------+----+-----------+

#----Остановил, бутсрап - /home/house": Permission denied
house@pgsql3:~$ sudo systemctl stop patroni
house@pgsql3:~$ sudo -u postgres patroni /etc/patroni.yml
2022-09-05 18:18:13,024 INFO: Selected new etcd server http://etcd2.ru-central1.internal:2379
2022-09-05 18:18:13,030 INFO: No PostgreSQL configuration items changed, nothing to reload.
2022-09-05 18:18:13,033 INFO: Lock owner: pgsql1; I am pgsql3
2022-09-05 18:18:13,037 INFO: trying to bootstrap from leader 'pgsql1'
could not change directory to "/home/house": Permission denied


--------------------------------------------------------------------------------------------------------------------------

#----Перешел в дирректорию, ошибка ушла при подключении: could not change directory to "/home/house": Permission denied

house@pgsql1:~$ cd ~postgres/
house@pgsql1:/var/lib/postgresql$ sudo -u postgres psql
psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
        Is the server running locally and accepting connections on that socket?
house@pgsql1:/var/lib/postgresql$ sudo -u postgres  psql -h localhost
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1))
Type "help" for help.

postgres=# create user replicator with password 'rep-pass_321';
CREATE ROLE
postgres=# \du
                                    List of roles
 Role name  |                         Attributes                         | Member of
------------+------------------------------------------------------------+-----------
 postgres   | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 replicator |                                                            | {}

postgres=#


#----Выдал роль Superuser для replicator

postgres=# ALTER USER replicator WITH SUPERUSER;
ALTER ROLE
postgres=# \du
                                    List of roles
 Role name  |                         Attributes                         | Member of
------------+------------------------------------------------------------+-----------
 postgres   | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 replicator | Superuser                                                  | {}

postgres=#


#----Проверил на 2ой ноде - запустилось

house@pgsql2:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni3 (7139975840175035628) -+----+-----------+
| pgsql1 | 10.129.0.23 | Leader  | running |  2 |           |
| pgsql2 | 10.129.0.35 | Replica | running |  2 |         0 |
+--------+-------------+---------+---------+----+-----------+


#----Проверил на 3ой ноде - запустилось
house@pgsql3:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni3 (7139975840175035628) -+----+-----------+
| pgsql1 | 10.129.0.23 | Leader  | running |  2 |           |
| pgsql2 | 10.129.0.35 | Replica | running |  2 |         0 |
| pgsql3 | 10.129.0.11 | Replica | running |  2 |         0 |
+--------+-------------+---------+---------+----+-----------+


#----установим pg_bouncer на каждом хосте с Патрони
sudo apt install -y pgbouncer


select usename,passwd from pg_shadow;
  usename   |                                                                passwd

------------+---------------------------------------------------------------------------------------------------------------------------------------
 postgres   | SCRAM-SHA-256$4096:KNl0g6hoyCheYR8jRppDQQ==$05dL6LFPXT0rBQwlZsIp2nZZdpM6NdMEWnhMX9BgrH0=:tDpfrqSk9Y+TvsYNTCevYg9g2ueUrC/tw6gneE1QRl8=
 replicator | SCRAM-SHA-256$4096:C8wsZN35l9Q82W0hPDl41g==$SbAfR+nUCNRSN2XWJ+dozn8vTyauHU32xwnpUuzsVG8=:HUSvIe/C0/qUgadMatQhA5CCcWyRNQBxNwiC5bAaPmo=
 admindb    | SCRAM-SHA-256$4096:/CYERE1hsAa/6qC1q/vLAA==$KgxxRlpF3sp1gZskAaCW3DYONPG+xQwaKvJ7U8u+OKo=:ul6zSC6npJ8OkYRUY8RR/3WCthR2Y+mFA1K3N4qdWNk=
(3 rows)


house@pgsql1:~$ sudo systemctl status pgbouncer
● pgbouncer.service - pgbouncer PostgreSQL connection pooler
     Loaded: loaded (/lib/systemd/system/pgbouncer.service; enabled; vendor preset: enabled)
     Active: active (running) since Mon 2022-09-05 20:54:55 UTC; 5s ago
   Main PID: 7790 (pgbouncer)
      Tasks: 2 (limit: 2236)
     Memory: 1.8M
        CPU: 7ms
     CGroup: /system.slice/pgbouncer.service
             └─7790 /usr/sbin/pgbouncer /etc/pgbouncer/pgbouncer.ini

Sep 05 20:54:55 pgsql1 systemd[1]: Starting pgbouncer PostgreSQL connection pooler...
Sep 05 20:54:55 pgsql1 pgbouncer[7790]: kernel file descriptor limit: 1024 (hard: 524288); max_client_conn: 100, max expected fd use: 172
Sep 05 20:54:55 pgsql1 pgbouncer[7790]: listening on 0.0.0.0:6432
Sep 05 20:54:55 pgsql1 pgbouncer[7790]: listening on [::]:6432
Sep 05 20:54:55 pgsql1 pgbouncer[7790]: listening on unix:/tmp/.s.PGSQL.6432
Sep 05 20:54:55 pgsql1 pgbouncer[7790]: process up: PgBouncer 1.17.0, libevent 2.1.12-stable (epoll), adns: c-ares 1.18.1, tls: OpenSSL 3.0.2 >
Sep 05 20:54:55 pgsql1 systemd[1]: Started pgbouncer PostgreSQL connection pooler.

house@pgsql1:~$ sudo -u postgres psql -p 6432 -h 127.0.0.1 otus
Password for user postgres: zalando_321
could not change directory to "/home/house": Permission denied
Password for user postgres:
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1))
Type "help" for help.

otus=# exit


house@pgsql1:~$ sudo -u postgres psql -p 6432 pgbouncer -h localhost
could not change directory to "/home/house": Permission denied
Password for user postgres:
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1), server 1.17.0/bouncer)
Type "help" for help.



#----На следующий день 1 и 2 ноды не стартанули.
house@pgsql1:~$ sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+----------+----+-----------+
| Member | Host        | Role    | State    | TL | Lag in MB |
+ Cluster: patroni3 (7139975840175035628) --+----+-----------+
| pgsql1 | 10.129.0.23 | Replica | starting |    |   unknown |
| pgsql2 | 10.129.0.35 | Replica | starting |    |   unknown |
| pgsql3 | 10.129.0.11 | Leader  | running  |  6 |           |
+--------+-------------+---------+----------+----+-----------+


#---- логи с патрони - не может подключиться к базе

#----1 нода
house@pgsql1:~$ sudo journalctl -eu patroni.service
Sep 06 19:01:38 pgsql1 patroni[1191]: 2022-09-06 19:01:38.188 UTC [1191] FATAL:  the database system is starting up
Sep 06 19:01:38 pgsql1 patroni[650]: 2022-09-06 19:01:38,189 WARNING: Retry got exception: 'connection problems'
Sep 06 19:01:38 pgsql1 patroni[650]: 2022-09-06 19:01:38,189 WARNING: Failed to determine PostgreSQL state from the connection, fa>Sep 06 19:01:38 pgsql1 patroni[650]: 2022-09-06 19:01:38,194 INFO: no action. I am (pgsql1), a secondary, and following a leader (>Sep 06 19:01:47 pgsql1 patroni[1193]: 2022-09-06 19:01:47.280 UTC [1193] FATAL:  the database system is starting up
Sep 06 19:01:47 pgsql1 patroni[1192]: localhost:5432 - rejecting connections
Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,281 INFO: Lock owner: pgsql3; I am pgsql1
Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,282 INFO: Still starting up as a standby.
Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,282 INFO: Lock owner: pgsql3; I am pgsql1
Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,282 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:01:47 pgsql1 patroni[1194]: 2022-09-06 19:01:47.284 UTC [1194] FATAL:  the database system is starting up
Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,835 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:01:47 pgsql1 patroni[1195]: 2022-09-06 19:01:47.836 UTC [1195] FATAL:  the database system is starting up
Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,837 WARNING: Retry got exception: 'connection problems'
Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,837 WARNING: Failed to determine PostgreSQL state from the connection, fa>Sep 06 19:01:47 pgsql1 patroni[650]: 2022-09-06 19:01:47,844 INFO: no action. I am (pgsql1), a secondary, and following a leader (>Sep 06 19:01:57 pgsql1 patroni[1197]: 2022-09-06 19:01:57.278 UTC [1197] FATAL:  the database system is starting up
Sep 06 19:01:57 pgsql1 patroni[1196]: localhost:5432 - rejecting connections
Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,280 INFO: Lock owner: pgsql3; I am pgsql1
Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,280 INFO: Still starting up as a standby.
Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,280 INFO: Lock owner: pgsql3; I am pgsql1
Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,280 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:01:57 pgsql1 patroni[1198]: 2022-09-06 19:01:57.282 UTC [1198] FATAL:  the database system is starting up
Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,843 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:01:57 pgsql1 patroni[1199]: 2022-09-06 19:01:57.844 UTC [1199] FATAL:  the database system is starting up
Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,845 WARNING: Retry got exception: 'connection problems'
Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,845 WARNING: Failed to determine PostgreSQL state from the connection, fa>Sep 06 19:01:57 pgsql1 patroni[650]: 2022-09-06 19:01:57,849 INFO: no action. I am (pgsql1), a secondary, and following a leader (


#----2 нода
house@pgsql2:~$ sudo journalctl -eu patroni.service
Sep 06 19:02:27 pgsql2 patroni[1102]: 2022-09-06 19:02:27.755 UTC [1102] FATAL:  the database system is starting up
Sep 06 19:02:27 pgsql2 patroni[652]: 2022-09-06 19:02:27,756 WARNING: Retry got exception: 'connection problems'
Sep 06 19:02:27 pgsql2 patroni[652]: 2022-09-06 19:02:27,756 WARNING: Failed to determine PostgreSQL state from the connection, fa>Sep 06 19:02:27 pgsql2 patroni[652]: 2022-09-06 19:02:27,760 INFO: no action. I am (pgsql2), a secondary, and following a leader (>Sep 06 19:02:37 pgsql2 patroni[1105]: 2022-09-06 19:02:37.280 UTC [1105] FATAL:  the database system is starting up
Sep 06 19:02:37 pgsql2 patroni[1104]: localhost:5432 - rejecting connections
Sep 06 19:02:37 pgsql2 patroni[652]: 2022-09-06 19:02:37,281 INFO: Lock owner: pgsql3; I am pgsql2
Sep 06 19:02:37 pgsql2 patroni[652]: 2022-09-06 19:02:37,281 INFO: Still starting up as a standby.
Sep 06 19:02:37 pgsql2 patroni[652]: 2022-09-06 19:02:37,282 INFO: Lock owner: pgsql3; I am pgsql2
Sep 06 19:02:37 pgsql2 patroni[652]: 2022-09-06 19:02:37,282 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:02:37 pgsql2 patroni[1106]: 2022-09-06 19:02:37.283 UTC [1106] FATAL:  the database system is starting up
Sep 06 19:02:38 pgsql2 patroni[652]: 2022-09-06 19:02:38,145 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:02:38 pgsql2 patroni[1109]: 2022-09-06 19:02:38.146 UTC [1109] FATAL:  the database system is starting up
Sep 06 19:02:38 pgsql2 patroni[652]: 2022-09-06 19:02:38,147 WARNING: Retry got exception: 'connection problems'
Sep 06 19:02:38 pgsql2 patroni[652]: 2022-09-06 19:02:38,147 WARNING: Failed to determine PostgreSQL state from the connection, fa>Sep 06 19:02:38 pgsql2 patroni[652]: 2022-09-06 19:02:38,151 INFO: no action. I am (pgsql2), a secondary, and following a leader (>Sep 06 19:02:47 pgsql2 patroni[1221]: 2022-09-06 19:02:47.279 UTC [1221] FATAL:  the database system is starting up
Sep 06 19:02:47 pgsql2 patroni[1220]: localhost:5432 - rejecting connections
Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,281 INFO: Lock owner: pgsql3; I am pgsql2
Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,281 INFO: Still starting up as a standby.
Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,281 INFO: Lock owner: pgsql3; I am pgsql2
Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,281 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:02:47 pgsql2 patroni[1222]: 2022-09-06 19:02:47.283 UTC [1222] FATAL:  the database system is starting up
Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,764 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 19:02:47 pgsql2 patroni[1223]: 2022-09-06 19:02:47.766 UTC [1223] FATAL:  the database system is starting up
Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,766 WARNING: Retry got exception: 'connection problems'
Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,766 WARNING: Failed to determine PostgreSQL state from the connection, fa>Sep 06 19:02:47 pgsql2 patroni[652]: 2022-09-06 19:02:47,771 INFO: no action. I am (pgsql2), a secondary, and following a leader 

#----3 нода - рабочая
house@pgsql3:~$ sudo journalctl -eu patroni.service
Sep 06 18:59:43 pgsql3 patroni[879]: localhost:5432 - accepting connections
Sep 06 18:59:43 pgsql3 patroni[650]: 2022-09-06 18:59:43,228 INFO: establishing a new patroni connection to the postgres cluster
Sep 06 18:59:45 pgsql3 patroni[650]: 2022-09-06 18:59:45,063 INFO: Got response from pgsql1 http://10.129.0.23:8008/patroni: {"sta>Sep 06 18:59:45 pgsql3 patroni[650]: 2022-09-06 18:59:45,063 WARNING: Request failed to pgsql1: GET http://10.129.0.23:8008/patron>Sep 06 18:59:45 pgsql3 patroni[650]: 2022-09-06 18:59:45,233 INFO: Got response from pgsql2 http://10.129.0.35:8008/patroni: {"sta>Sep 06 18:59:45 pgsql3 patroni[650]: 2022-09-06 18:59:45,233 WARNING: Request failed to pgsql2: GET http://10.129.0.35:8008/patron>Sep 06 18:59:45 pgsql3 patroni[650]: 2022-09-06 18:59:45,239 WARNING: Could not activate Linux watchdog device: "Can't open watchd>Sep 06 18:59:45 pgsql3 patroni[650]: 2022-09-06 18:59:45,248 INFO: promoted self to leader by acquiring session lock
Sep 06 18:59:45 pgsql3 patroni[873]: 2022-09-06 18:59:45.251 UTC [873] LOG:  received promote request
Sep 06 18:59:45 pgsql3 patroni[873]: 2022-09-06 18:59:45.251 UTC [873] LOG:  redo is not required
Sep 06 18:59:45 pgsql3 patroni[889]: server promoting
Sep 06 18:59:45 pgsql3 patroni[650]: 2022-09-06 18:59:45,252 INFO: cleared rewind state after becoming the leader
Sep 06 18:59:45 pgsql3 patroni[873]: 2022-09-06 18:59:45.264 UTC [873] LOG:  selected new timeline ID: 6
Sep 06 18:59:46 pgsql3 patroni[873]: 2022-09-06 18:59:46.731 UTC [873] LOG:  archive recovery complete
Sep 06 18:59:46 pgsql3 patroni[869]: 2022-09-06 18:59:46.785 UTC [869] LOG:  database system is ready to accept connections
Sep 06 18:59:47 pgsql3 patroni[650]: 2022-09-06 18:59:47,300 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 18:59:57 pgsql3 patroni[650]: 2022-09-06 18:59:57,272 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:00:07 pgsql3 patroni[650]: 2022-09-06 19:00:07,278 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:00:17 pgsql3 patroni[650]: 2022-09-06 19:00:17,267 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:00:27 pgsql3 patroni[650]: 2022-09-06 19:00:27,268 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:00:37 pgsql3 patroni[650]: 2022-09-06 19:00:37,266 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:00:47 pgsql3 patroni[650]: 2022-09-06 19:00:47,266 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:00:57 pgsql3 patroni[650]: 2022-09-06 19:00:57,266 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:01:07 pgsql3 patroni[650]: 2022-09-06 19:01:07,266 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:01:17 pgsql3 patroni[650]: 2022-09-06 19:01:17,266 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:01:27 pgsql3 patroni[650]: 2022-09-06 19:01:27,267 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:01:37 pgsql3 patroni[650]: 2022-09-06 19:01:37,279 INFO: no action. I am (pgsql3), the leader with the lock
Sep 06 19:01:47 pgsql3 patroni[650]: 2022-09-06 19:01:47,275 INFO: no action. I am (pgsql3), the leader with the lock


#----на 3 ноды проверил доступность баз - т.е. проблема коннектов получается...

house@pgsql3:~$ sudo -u postgres psql -p 6432 pgbouncer -h localhost
could not change directory to "/home/house": Permission denied
Password for user postgres:
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1), server 1.17.0/bouncer)
Type "help" for help.

pgbouncer=# show clients;

 type |   user   | database  | state  | addr | port  | local_addr | local_port |      connect_time       |      request_time       | wait | wait_us | close_needed |      ptr       | link | remote_pid | tls
------+----------+-----------+--------+------+-------+------------+------------+-------------------------+-------------------------+------+---------+--------------+----------------+------+------------+-----
 C    | postgres | pgbouncer | active | ::1  | 38680 | ::1        |       6432 | 2022-09-06 21:17:53 UTC | 2022-09-06 21:18:54 UTC |   39 |  423053 |            0 | 0x55d80ba4ae00 |      |          0 |
(1 row)


house@pgsql3:~$ sudo -u postgres psql -h localhost
could not change directory to "/home/house": Permission denied
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1))
Type "help" for help.

postgres=# \du
                                    List of roles
 Role name  |                         Attributes                         | Member of
------------+------------------------------------------------------------+-----------
 admindb    |                                                            | {}
 postgres   | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 replicator | Superuser                                                  | {}


postgres=# \l
                                  List of databases
   Name    |  Owner   | Encoding |   Collate   |    Ctype    |   Access privileges
-----------+----------+----------+-------------+-------------+-----------------------
 otus      | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 postgres  | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 template0 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
 template1 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
(4 rows)

postgres=#








------------------------------------------------------------------------------------------------------------------------------


postgres=# ALTER USER replicator WITH REPLICATION;
ALTER ROLE
postgres=# \du
                                    List of roles
 Role name  |                         Attributes                         | Member of
------------+------------------------------------------------------------+-----------
 admindb    |                                                            | {}
 postgres   | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 replicator | Superuser, Replication                                     | {}





create user admin with password 'admin_321';


postgres=# create user admin with password 'admin_321';
CREATE ROLE
postgres=# ALTER USER admin WITH createrole;
ALTER ROLE
postgres=# ALTER USER admin WITH createdb;
ALTER ROLE
postgres=# create user rewind_user with password 'rewind_password_321';
CREATE ROLE
postgres=# \du
                                    List of roles
  Role name  |                         Attributes                         | Member of
-------------+------------------------------------------------------------+-----------
 admin       | Create role, Create DB                                     | {}
 admindb     |                                                            | {}
 postgres    | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 replicator  | Superuser, Replication                                     | {}
 rewind_user |                                                            | {}


#----От пользователя house доступ не получил, только от рута.

root@pgsql3:/var/lib/postgresql/14/main# cat pg_hba.conf
-------------+---------------------------------------------------------
# TYPE  DATABASE        USER            ADDRESS                 METHOD

# "local" is for Unix domain socket connections only
local   all             all                                     trust
# IPv4 local connections:
host    all             all             127.0.0.1/32            trust
# IPv6 local connections:
host    all             all             ::1/128                 trust
# Allow replication connections from localhost, by a user with the
# replication privilege.
local   replication     all                                     trust
host    replication     all             127.0.0.1/32            trust
host    replication     all             ::1/128                 trust

host replication replicator 10.0.0.0/8 md5
host all all 10.0.0.0/8 md5
root@pgsql3:/var/lib/postgresql/14/main#
-------------+---------------------------------------------------------


#---- Останавливаю патрони на 3х нодах и бустрапою

root@pgsql3:~# sudo systemctl stop patroni
root@pgsql3:~# sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni3 (7139975840175035628) -+----+-----------+
| pgsql1 | 10.129.0.23 | Replica | stopped |    |   unknown |
| pgsql2 | 10.129.0.35 | Replica | stopped |    |   unknown |
| pgsql3 | 10.129.0.11 | Replica | stopped |    |   unknown |
+--------+-------------+---------+---------+----+-----------+

root@pgsql3:~# sudo -u postgres patroni /etc/patroni.yml
2022-09-07 13:53:04,320 INFO: Selected new etcd server http://etcd2.ru-central1.internal:2379
2022-09-07 13:53:04,325 INFO: No PostgreSQL configuration items changed, nothing to reload.
could not change directory to "/root": Permission denied
could not change directory to "/root": Permission denied
2022-09-07 13:53:04,334 WARNING: Postgresql is not running.
could not change directory to "/root": Permission denied
2022-09-07 13:53:04,334 INFO: Lock owner: None; I am pgsql3
2022-09-07 13:53:04,336 INFO: pg_controldata:
  pg_control version number: 1300
  Catalog version number: 202107181
  Database system identifier: 7139975840175035628
  Database cluster state: shut down
  pg_control last modified: Wed Sep  7 13:52:09 2022
  Latest checkpoint location: 0/6006600
  Latest checkpoint's REDO location: 0/6006600
  Latest checkpoint's REDO WAL file: 0000000B0000000000000006
  Latest checkpoint's TimeLineID: 11
  Latest checkpoint's PrevTimeLineID: 11
  Latest checkpoint's full_page_writes: on
  Latest checkpoint's NextXID: 0:743
  Latest checkpoint's NextOID: 24580
  Latest checkpoint's NextMultiXactId: 1
  Latest checkpoint's NextMultiOffset: 0
  Latest checkpoint's oldestXID: 727
  Latest checkpoint's oldestXID's DB: 1
  Latest checkpoint's oldestActiveXID: 0
  Latest checkpoint's oldestMultiXid: 1
  Latest checkpoint's oldestMulti's DB: 1
  Latest checkpoint's oldestCommitTsXid: 0
  Latest checkpoint's newestCommitTsXid: 0
  Time of latest checkpoint: Wed Sep  7 13:52:09 2022
  Fake LSN counter for unlogged rels: 0/3E8
  Minimum recovery ending location: 0/0
  Min recovery ending loc's timeline: 0
  Backup start location: 0/0
  Backup end location: 0/0
  End-of-backup record required: no
  wal_level setting: replica
  wal_log_hints setting: on
  max_connections setting: 100
  max_worker_processes setting: 8
  max_wal_senders setting: 10
  max_prepared_xacts setting: 0
  max_locks_per_xact setting: 64
  track_commit_timestamp setting: off
  Maximum data alignment: 8
  Database block size: 8192
  Blocks per segment of large relation: 131072
  WAL block size: 8192
  Bytes per WAL segment: 16777216
  Maximum length of identifiers: 64
  Maximum columns in an index: 32
  Maximum size of a TOAST chunk: 1996
  Size of a large-object chunk: 2048
  Date/time type storage: 64-bit integers
  Float8 argument passing: by value
  Data page checksum version: 1
  Mock authentication nonce: 70bf07cd3737798d1ee5de903ee8d320f099c0d837509f75413ec2e22cba1c58

could not change directory to "/root": Permission denied
2022-09-07 13:53:04,341 INFO: Lock owner: None; I am pgsql3
2022-09-07 13:53:04,345 INFO: starting as a secondary
could not change directory to "/root": Permission denied
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/usr/lib/python3.10/multiprocessing/spawn.py", line 225, in prepare
    os.chdir(data['dir'])
PermissionError: [Errno 13] Permission denied: '/root'
2022-09-07 13:53:14,329 INFO: Lock owner: None; I am pgsql3
2022-09-07 13:53:14,330 INFO: not healthy enough for leader race
2022-09-07 13:53:14,333 INFO: restarting after failure in progress
2022-09-07 13:53:24,329 INFO: Lock owner: None; I am pgsql3
2022-09-07 13:53:24,329 INFO: not healthy enough for leader race
2022-09-07 13:53:24,333 INFO: restarting after failure in progress
^C^CException ignored in: <module 'threading' from '/usr/lib/python3.10/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1560, in _shutdown
    lock.acquire()
KeyboardInterrupt:


#----Безуспешно

root@pgsql3:~# sudo systemctl start patroni

root@pgsql3:~# sudo patronictl -c /etc/patroni.yml list
+--------+-------------+---------+----------+----+-----------+
| Member | Host        | Role    | State    | TL | Lag in MB |
+ Cluster: patroni3 (7139975840175035628) --+----+-----------+
| pgsql1 | 10.129.0.23 | Replica | starting |    |   unknown |
| pgsql2 | 10.129.0.35 | Replica | starting |    |   unknown |
| pgsql3 | 10.129.0.11 | Leader  | running  | 12 |           |
+--------+-------------+---------+----------+----+-----------+


#----Проверил доступность с 1 и 2 ноды
house@pgsql1:~$ ping pgsql2
PING pgsql2.ru-central1.internal (10.129.0.35) 56(84) bytes of data.
64 bytes from pgsql2.ru-central1.internal (10.129.0.35): icmp_seq=1 ttl=63 time=0.802 ms
64 bytes from pgsql2.ru-central1.internal (10.129.0.35): icmp_seq=2 ttl=63 time=0.320 ms
64 bytes from pgsql2.ru-central1.internal (10.129.0.35): icmp_seq=3 ttl=63 time=0.325 ms
^C
--- pgsql2.ru-central1.internal ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2005ms
rtt min/avg/max/mdev = 0.320/0.482/0.802/0.226 ms
house@pgsql1:~$ ping pgsql3
PING pgsql3.ru-central1.internal (10.129.0.11) 56(84) bytes of data.
64 bytes from pgsql3.ru-central1.internal (10.129.0.11): icmp_seq=1 ttl=63 time=0.818 ms
64 bytes from pgsql3.ru-central1.internal (10.129.0.11): icmp_seq=2 ttl=63 time=0.333 ms
^C
--- pgsql3.ru-central1.internal ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 0.333/0.575/0.818/0.242 ms


#---- логи с патрони

#----1 нода 
root@pgsql1:/home/house# sudo journalctl -eu patroni.service
Sep 07 13:59:25 pgsql1 patroni[3281]: 2022-09-07 13:59:25,064 INFO: Lock owner: pgsql3; I am pgsql1
Sep 07 13:59:25 pgsql1 patroni[3281]: 2022-09-07 13:59:25,064 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:25 pgsql1 patroni[3486]: 2022-09-07 13:59:25.065 UTC [3486] FATAL:  the database system is starting up
Sep 07 13:59:25 pgsql1 patroni[3281]: 2022-09-07 13:59:25,917 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:25 pgsql1 patroni[3487]: 2022-09-07 13:59:25.919 UTC [3487] FATAL:  the database system is starting up
Sep 07 13:59:25 pgsql1 patroni[3281]: 2022-09-07 13:59:25,919 WARNING: Retry got exception: 'connection problems'
Sep 07 13:59:25 pgsql1 patroni[3281]: 2022-09-07 13:59:25,920 WARNING: Failed to determine PostgreSQL state from the connection, f>Sep 07 13:59:25 pgsql1 patroni[3281]: 2022-09-07 13:59:25,925 INFO: no action. I am (pgsql1), a secondary, and following a leader >Sep 07 13:59:35 pgsql1 patroni[3491]: 2022-09-07 13:59:35.062 UTC [3491] FATAL:  the database system is starting up
Sep 07 13:59:35 pgsql1 patroni[3490]: localhost:5432 - rejecting connections
Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,063 INFO: Lock owner: pgsql3; I am pgsql1
Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,063 INFO: Still starting up as a standby.
Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,064 INFO: Lock owner: pgsql3; I am pgsql1
Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,064 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:35 pgsql1 patroni[3492]: 2022-09-07 13:59:35.066 UTC [3492] FATAL:  the database system is starting up
Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,587 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:35 pgsql1 patroni[3493]: 2022-09-07 13:59:35.589 UTC [3493] FATAL:  the database system is starting up
Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,589 WARNING: Retry got exception: 'connection problems'
Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,589 WARNING: Failed to determine PostgreSQL state from the connection, f>Sep 07 13:59:35 pgsql1 patroni[3281]: 2022-09-07 13:59:35,593 INFO: no action. I am (pgsql1), a secondary, and following a leader >Sep 07 13:59:45 pgsql1 patroni[3498]: 2022-09-07 13:59:45.062 UTC [3498] FATAL:  the database system is starting up
Sep 07 13:59:45 pgsql1 patroni[3497]: localhost:5432 - rejecting connections
Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,064 INFO: Lock owner: pgsql3; I am pgsql1
Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,064 INFO: Still starting up as a standby.
Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,064 INFO: Lock owner: pgsql3; I am pgsql1
Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,064 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:45 pgsql1 patroni[3499]: 2022-09-07 13:59:45.066 UTC [3499] FATAL:  the database system is starting up
Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,927 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:45 pgsql1 patroni[3500]: 2022-09-07 13:59:45.929 UTC [3500] FATAL:  the database system is starting up
Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,929 WARNING: Retry got exception: 'connection problems'
Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,929 WARNING: Failed to determine PostgreSQL state from the connection, f>Sep 07 13:59:45 pgsql1 patroni[3281]: 2022-09-07 13:59:45,934 INFO: no action. I am (pgsql1), a secondary, and following a leader 

#----2 нода 
house@pgsql2:~$ sudo journalctl -eu patroni.service
Sep 07 13:59:05 pgsql2 patroni[3729]: 2022-09-07 13:59:05,845 INFO: no action. I am (pgsql2), a secondary, and following a leader >Sep 07 13:59:15 pgsql2 patroni[4000]: 2022-09-07 13:59:15.062 UTC [4000] FATAL:  the database system is starting up
Sep 07 13:59:15 pgsql2 patroni[3999]: localhost:5432 - rejecting connections
Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,064 INFO: Lock owner: pgsql3; I am pgsql2
Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,064 INFO: Still starting up as a standby.
Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,065 INFO: Lock owner: pgsql3; I am pgsql2
Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,065 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:15 pgsql2 patroni[4001]: 2022-09-07 13:59:15.066 UTC [4001] FATAL:  the database system is starting up
Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,787 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:15 pgsql2 patroni[4002]: 2022-09-07 13:59:15.789 UTC [4002] FATAL:  the database system is starting up
Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,790 WARNING: Retry got exception: 'connection problems'
Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,790 WARNING: Failed to determine PostgreSQL state from the connection, f>Sep 07 13:59:15 pgsql2 patroni[3729]: 2022-09-07 13:59:15,797 INFO: no action. I am (pgsql2), a secondary, and following a leader >Sep 07 13:59:25 pgsql2 patroni[4006]: 2022-09-07 13:59:25.062 UTC [4006] FATAL:  the database system is starting up
Sep 07 13:59:25 pgsql2 patroni[4005]: localhost:5432 - rejecting connections
Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,064 INFO: Lock owner: pgsql3; I am pgsql2
Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,064 INFO: Still starting up as a standby.
Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,065 INFO: Lock owner: pgsql3; I am pgsql2
Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,065 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:25 pgsql2 patroni[4007]: 2022-09-07 13:59:25.066 UTC [4007] FATAL:  the database system is starting up
Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,287 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:25 pgsql2 patroni[4008]: 2022-09-07 13:59:25.289 UTC [4008] FATAL:  the database system is starting up
Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,289 WARNING: Retry got exception: 'connection problems'
Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,290 WARNING: Failed to determine PostgreSQL state from the connection, f>Sep 07 13:59:25 pgsql2 patroni[3729]: 2022-09-07 13:59:25,297 INFO: no action. I am (pgsql2), a secondary, and following a leader >Sep 07 13:59:35 pgsql2 patroni[4012]: 2022-09-07 13:59:35.063 UTC [4012] FATAL:  the database system is starting up
Sep 07 13:59:35 pgsql2 patroni[4011]: localhost:5432 - rejecting connections
Sep 07 13:59:35 pgsql2 patroni[3729]: 2022-09-07 13:59:35,064 INFO: Lock owner: pgsql3; I am pgsql2
Sep 07 13:59:35 pgsql2 patroni[3729]: 2022-09-07 13:59:35,064 INFO: Still starting up as a standby.
Sep 07 13:59:35 pgsql2 patroni[3729]: 2022-09-07 13:59:35,065 INFO: Lock owner: pgsql3; I am pgsql2
Sep 07 13:59:35 pgsql2 patroni[3729]: 2022-09-07 13:59:35,065 INFO: establishing a new patroni connection to the postgres cluster
Sep 07 13:59:35 pgsql2 patroni[4013]: 2022-09-07 13:59:35.066 UTC [4013] FATAL:  the database system is starting up


#----3 нода 
root@pgsql3:~# sudo journalctl -eu patroni.service
Sep 07 13:54:45 pgsql3 patroni[1859]: 2022-09-07 13:54:45,134 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:54:55 pgsql3 patroni[1859]: 2022-09-07 13:54:55,051 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:55:05 pgsql3 patroni[1859]: 2022-09-07 13:55:05,051 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:55:11 pgsql3 patroni[1906]: 2022-09-07 13:55:11.219 UTC [1906] LOG:  could not send data to client: Connection reset by >Sep 07 13:55:11 pgsql3 patroni[1906]: 2022-09-07 13:55:11.219 UTC [1906] STATEMENT:  START_REPLICATION SLOT "pgsql2" 0/5000000 TIM>Sep 07 13:55:15 pgsql3 patroni[1859]: 2022-09-07 13:55:15,052 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:55:25 pgsql3 patroni[1859]: 2022-09-07 13:55:25,050 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:55:35 pgsql3 patroni[1859]: 2022-09-07 13:55:35,098 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:55:45 pgsql3 patroni[1859]: 2022-09-07 13:55:45,049 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:55:55 pgsql3 patroni[1859]: 2022-09-07 13:55:55,048 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:56:05 pgsql3 patroni[1859]: 2022-09-07 13:56:05,051 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:56:15 pgsql3 patroni[1859]: 2022-09-07 13:56:15,053 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:56:25 pgsql3 patroni[1859]: 2022-09-07 13:56:25,051 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:56:35 pgsql3 patroni[1859]: 2022-09-07 13:56:35,052 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:56:45 pgsql3 patroni[1859]: 2022-09-07 13:56:45,051 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:56:55 pgsql3 patroni[1859]: 2022-09-07 13:56:55,048 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:57:05 pgsql3 patroni[1859]: 2022-09-07 13:57:05,048 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:57:14 pgsql3 patroni[1915]: 2022-09-07 13:57:14.853 UTC [1915] LOG:  could not send data to client: Connection reset by >Sep 07 13:57:14 pgsql3 patroni[1915]: 2022-09-07 13:57:14.853 UTC [1915] STATEMENT:  START_REPLICATION SLOT "pgsql1" 0/5000000 TIM>Sep 07 13:57:15 pgsql3 patroni[1859]: 2022-09-07 13:57:15,054 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:57:25 pgsql3 patroni[1859]: 2022-09-07 13:57:25,048 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:57:35 pgsql3 patroni[1859]: 2022-09-07 13:57:35,051 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:57:45 pgsql3 patroni[1859]: 2022-09-07 13:57:45,054 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:57:55 pgsql3 patroni[1859]: 2022-09-07 13:57:55,049 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:58:05 pgsql3 patroni[1859]: 2022-09-07 13:58:05,047 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:58:15 pgsql3 patroni[1859]: 2022-09-07 13:58:15,049 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:58:25 pgsql3 patroni[1859]: 2022-09-07 13:58:25,051 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:58:35 pgsql3 patroni[1859]: 2022-09-07 13:58:35,049 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:58:45 pgsql3 patroni[1859]: 2022-09-07 13:58:45,055 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:58:55 pgsql3 patroni[1859]: 2022-09-07 13:58:55,052 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:59:05 pgsql3 patroni[1859]: 2022-09-07 13:59:05,050 INFO: no action. I am (pgsql3), the leader with the lock
Sep 07 13:59:15 pgsql3 patroni[1859]: 2022-09-07 13:59:15,049 INFO: no action. I am (pgsql3), the leader with the lock


-------------------------------------------------------------------------------------------------------------------------------------------------------------
#----Для решения данной проблемы: стопаем сервисы постгре и патрони
systemctl stop patroni
systemctl stop postgresql


#----Удаляем каталог /var/lib/postgresql/14
cd /var/lib/postgresql
rm -rf 14

#----бутстрапим патрони

root@pgsql1:/home/house# patronictl -c /etc/patroni.yml list
+--------+-------------+---------+---------+----+-----------+
| Member | Host        | Role    | State   | TL | Lag in MB |
+ Cluster: patroni3 (7139975840175035628) -+----+-----------+
| pgsql1 | 10.129.0.23 | Leader  | running | 18 |           |
| pgsql2 | 10.129.0.35 | Replica | running | 18 |         0 |
| pgsql3 | 10.129.0.11 | Replica | running | 18 |         0 |
+--------+-------------+---------+---------+----+-----------+


#----Подключился к pgbouncer
root@pgsql1:/home/house# sudo -u postgres psql -p 6432 pgbouncer -h localhost
could not change directory to "/home/house": Permission denied
Password for user postgres:
psql (14.5 (Ubuntu 14.5-1.pgdg22.04+1), server 1.17.0/bouncer)
Type "help" for help.

pgbouncer=# show clients;

 type |   user   | database  | state  | addr | port  | local_addr | local_port |      connect_time       |      request_time       | wait | wait_us | close_needed |      ptr       | link | remote_pid | tls
------+----------+-----------+--------+------+-------+------------+------------+-------------------------+-------------------------+------+---------+--------------+----------------+------+------------+-----
 C    | postgres | pgbouncer | active | ::1  | 36364 | ::1        |       6432 | 2022-09-12 09:51:06 UTC | 2022-09-12 09:51:58 UTC |   22 |  245429 |            0 | 0x561bbc39ce00 |      |          0 |
(1 row)


#----Нагрузил базу локально
root@pgsql1:/home/house# sudo -u postgres pgbench -p 6432 -i -d otus -h localhost
Password:
dropping old tables...
NOTICE:  table "pgbench_accounts" does not exist, skipping
NOTICE:  table "pgbench_branches" does not exist, skipping
NOTICE:  table "pgbench_history" does not exist, skipping
NOTICE:  table "pgbench_tellers" does not exist, skipping
creating tables...
generating data (client-side)...
100000 of 100000 tuples (100%) done (elapsed 0.03 s, remaining 0.00 s)
vacuuming...
creating primary keys...
done in 1.10 s (drop tables 0.01 s, create tables 0.06 s, client-side generate 0.64 s, vacuum 0.09 s, primary keys 0.30 s).
Type "help" for help.

#----Проверил статисктику
pgbouncer=# SHOW STATS_TOTALS;
 database  | xact_count | query_count | bytes_received | bytes_sent | xact_time | query_time | wait_time
-----------+------------+-------------+----------------+------------+-----------+------------+-----------
 otus      |         17 |          31 |        1590793 |       2170 |   1112172 |    1111522 |     43271
 pgbouncer |          5 |           5 |              0 |          0 |         0 |          0 |         0
(2 rows)



#----Нагрузил с  2ой ноды 
root@pgsql2:/home/house# sudo -u postgres pgbench -p 6432 -i -d otus -h 10.129.0.23
Password:
dropping old tables...
creating tables...
generating data (client-side)...
100000 of 100000 tuples (100%) done (elapsed 0.02 s, remaining 0.00 s)
vacuuming...
creating primary keys...
done in 2.02 s (drop tables 0.01 s, create tables 0.02 s, client-side generate 1.67 s, vacuum 0.06 s, primary keys 0.25 s).


#----Проверил - статистика увеличилась.
pgbouncer=# SHOW STATS_TOTALS;
 database  | xact_count | query_count | bytes_received | bytes_sent | xact_time | query_time | wait_time
-----------+------------+-------------+----------------+------------+-----------+------------+-----------
 otus      |         43 |          85 |        4772035 |       3686 |   4083100 |    4079193 |     43781
 pgbouncer |          6 |           6 |              0 |          0 |         0 |          0 |         0
(2 rows)


 type |   user   | database | state |   addr    | port | local_addr | local_port |      connect_time       |      request_time       | wait | wait_us | close_needed |      ptr       | link | remote_pid | tls
------+----------+----------+-------+-----------+------+------------+------------+-------------------------+-------------------------+------+---------+--------------+----------------+------+------------+-----
 S    | postgres | otus     | used  | 127.0.0.1 | 5432 | 127.0.0.1  |      36534 | 2022-09-12 11:20:47 UTC | 2022-09-12 11:26:54 UTC |    0 |       0 |            0 | 0x561bbc3a9660 |      |       1741 |
(1 row)



#---- развернем 2 ВМ для HAProxy
sudo apt install -y --no-install-recommends software-properties-common && sudo add-apt-repository -y ppa:vbernat/haproxy-2.5 && sudo apt install -y haproxy=2.5.\*



#----Проверим ответ мастера
curl -v 10.129.0.23:8008/master

house@proxy1:~$ curl -v 10.129.0.23:8008/master
*   Trying 10.129.0.23:8008...
* Connected to 10.129.0.23 (10.129.0.23) port 8008 (#0)
> GET /master HTTP/1.1
> Host: 10.129.0.23:8008
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Server: BaseHTTP/0.6 Python/3.10.4
< Date: Mon, 12 Sep 2022 12:35:58 GMT
< Content-Type: application/json
<
* Closing connection 0
{"state": "running", "postmaster_start_time": "2022-09-12 09:26:12.524834+00:00", "role": "master", "server_version": 140005, "xlog": {"location": 274612880}, "timeline": 19, "replication": [{"usename": "replicator", "application_name": "pgsql3", "client_addr": "10.129.0.11", "state": "streaming", "sync_state": "async", "sync_priority": 0}, {"usename": "replicator", "application_name": "pgsql2", "client_addr": "10.129.0.35", "state": "streaming", "sync_state": "async", "sync_priority": 0}], "dcs_last_seen": 1662986155, "database_system_identifier": "7139975840175035628", "patroni": {"version": "2.1.4", "scope": "patroni3"}}house@proxy1:~$


sudo apt update && sudo apt upgrade -y && sudo apt install -y postgresql-client-common && sudo apt install postgresql-client -y

#zalando_321

psql -p 6432 -d otus -h 10.129.0.23 -U postgres

#----Проверим подключение
house@proxy1:~$ psql -p 6432 -d otus -h 10.129.0.23 -U postgres
Password for user postgres:
psql (14.5 (Ubuntu 14.5-0ubuntu0.22.04.1))
Type "help" for help.

otus=# 

otus=# \l
                                  List of databases
   Name    |  Owner   | Encoding |   Collate   |    Ctype    |   Access privileges
-----------+----------+----------+-------------+-------------+-----------------------
 otus      | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 postgres  | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 template0 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
 template1 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
(4 rows)

otus=# create database haproxy;
CREATE DATABASE
otus=# \l
                                  List of databases
   Name    |  Owner   | Encoding |   Collate   |    Ctype    |   Access privileges
-----------+----------+----------+-------------+-------------+-----------------------
 haproxy   | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 otus      | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 postgres  | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 template0 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
 template1 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
(5 rows)



udo nano /etc/haproxy/haproxy.cfg
listen postgres_write
    bind *:5432
    mode            tcp
    option httpchk
    http-check connect
    http-check send meth GET uri /master
    http-check expect status 200
    default-server inter 10s fall 3 rise 3 on-marked-down shutdown-sessions
    server pgsql1 10.129.0.23:6432 check port 8008
    server pgsql2 10.129.0.35:6432 check port 8008
    server pgsql3 10.129.0.11:6432 check port 8008

listen postgres_read
    bind *:5433
    mode            tcp
    http-check connect
    http-check send meth GET uri /replica
    http-check expect status 200
    default-server inter 10s fall 3 rise 3 on-marked-down shutdown-sessions
    server pgsql1 10.129.0.23:6432 check port 8008
    server pgsql2 10.129.0.35:6432 check port 8008
    server pgsql3 10.129.0.11:6432 check port 8008

	
#----	Рестартуем хапрокси
house@proxy1:~$ sudo systemctl restart haproxy.service
house@proxy1:~$ sudo systemctl status haproxy.service
● haproxy.service - HAProxy Load Balancer
     Loaded: loaded (/lib/systemd/system/haproxy.service; enabled; vendor preset: enabled)
     Active: active (running) since Mon 2022-09-12 14:30:47 UTC; 23s ago
       Docs: man:haproxy(1)
             file:/usr/share/doc/haproxy/configuration.txt.gz
   Main PID: 24652 (haproxy)
      Tasks: 3 (limit: 2236)
     Memory: 38.2M
        CPU: 58ms
     CGroup: /system.slice/haproxy.service
             ├─24652 /usr/sbin/haproxy -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/hapr>             └─24654 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock

Sep 12 14:30:47 proxy1 haproxy[24652]: [WARNING]  (24652) : config : proxy 'postgres_read' uses http-check rules without 'option h>Sep 12 14:30:47 proxy1 haproxy[24652]: [NOTICE]   (24652) : New worker (24654) forked
Sep 12 14:30:47 proxy1 haproxy[24652]: [NOTICE]   (24652) : Loading success.
Sep 12 14:30:47 proxy1 systemd[1]: Started HAProxy Load Balancer.
Sep 12 14:30:48 proxy1 haproxy[24654]: [WARNING]  (24654) : Server postgres_write/pgsql2 is DOWN, reason: Layer7 wrong status, cod>Sep 12 14:30:48 proxy1 haproxy[24654]: Server postgres_write/pgsql2 is DOWN, reason: Layer7 wrong status, code: 503, info: "Servic>Sep 12 14:30:48 proxy1 haproxy[24654]: Server postgres_write/pgsql2 is DOWN, reason: Layer7 wrong status, code: 503, info: "Servic>Sep 12 14:30:50 proxy1 haproxy[24654]: [WARNING]  (24654) : Server postgres_write/pgsql3 is DOWN, reason: Layer7 wrong status, cod>Sep 12 14:30:50 proxy1 haproxy[24654]: Server postgres_write/pgsql3 is DOWN, reason: Layer7 wrong status, code: 503, info: "Servic>Sep 12 14:30:50 proxy1 haproxy[24654]: Server postgres_write/pgsql3 is DOWN, reason: Layer7 wrong status, code: 503, info: "Servic>


#---- Логи
Sep 12 14:30:47 proxy1 haproxy[1924]: [NOTICE]   (1924) : haproxy version is 2.5.8-1ppa1~jammy
Sep 12 14:30:47 proxy1 haproxy[1924]: [NOTICE]   (1924) : path to executable is /usr/sbin/haproxy
Sep 12 14:30:47 proxy1 haproxy[1924]: [WARNING]  (1924) : Exiting Master process...
Sep 12 14:30:47 proxy1 haproxy[1924]: [ALERT]    (1924) : Current worker (1926) exited with code 143 (Terminated)
Sep 12 14:30:47 proxy1 haproxy[1924]: [WARNING]  (1924) : All workers exited. Exiting... (0)
Sep 12 14:30:47 proxy1 haproxy[24652]: [NOTICE]   (24652) : haproxy version is 2.5.8-1ppa1~jammy
Sep 12 14:30:47 proxy1 haproxy[24652]: [NOTICE]   (24652) : path to executable is /usr/sbin/haproxy
Sep 12 14:30:47 proxy1 haproxy[24652]: [WARNING]  (24652) : config : parsing [/etc/haproxy/haproxy.cfg:23] : 'option httplog' not usable with proxy 'postgres_write' (needs 'mode http'). Falling back to 'option tcplog'.
Sep 12 14:30:47 proxy1 haproxy[24652]: [WARNING]  (24652) : config : parsing [/etc/haproxy/haproxy.cfg:23] : 'option httplog' not usable with proxy 'postgres_read' (needs 'mode http'). Falling back to 'option tcplog'.
Sep 12 14:30:47 proxy1 haproxy[24652]: [WARNING]  (24652) : config : proxy 'postgres_read' uses http-check rules without 'option httpchk', so the rules are ignored.
Sep 12 14:30:47 proxy1 haproxy[24652]: [NOTICE]   (24652) : New worker (24654) forked
Sep 12 14:30:47 proxy1 haproxy[24652]: [NOTICE]   (24652) : Loading success.
Sep 12 14:30:48 proxy1 haproxy[24654]: [WARNING]  (24654) : Server postgres_write/pgsql2 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Sep 12 14:30:48 proxy1 haproxy[24654]: Server postgres_write/pgsql2 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Sep 12 14:30:48 proxy1 haproxy[24654]: Server postgres_write/pgsql2 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Sep 12 14:30:50 proxy1 haproxy[24654]: [WARNING]  (24654) : Server postgres_write/pgsql3 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Sep 12 14:30:50 proxy1 haproxy[24654]: Server postgres_write/pgsql3 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Sep 12 14:30:50 proxy1 haproxy[24654]: Server postgres_write/pgsql3 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.      

#----Проверим подключение.
house@proxy1:~$ psql -h localhost -d otus -U postgres -p 5432
Password for user postgres:
psql (14.5 (Ubuntu 14.5-0ubuntu0.22.04.1))
Type "help" for help.

otus=# \q
